Bypass Apply Requirements with --f

pcalley opened the pull request:
Adds the -f or --force option to atlantis apply comment commands:
```  -f, --force                 Force Atlantis to ignore apply requirments.```

Dynamic config related PR: https://github.com/lyft/atlantis-dynamic-config/pull/8

This will be disabled when first deployed.

pcalley commented:
ptal #orchestration @nishkrishnan @msarvar @Aayyush

pcalley commented:
/ptal #orchestration @nishkrishnan @msarvar @Aayyush

msarvar commented on server/events/project_command_context_builder.go:
```diff
@@ -37,7 +37,7 @@ type ProjectCommandContextBuilder interface {
 		commentFlags []string,
 		repoDir string,
-		automerge, deleteSourceBranchOnMerge, parallelApply, parallelPlan, verbose bool,
+		automerge, deleteSourceBranchOnMerge, parallelApply, parallelPlan, verbose, force bool,
```
I think these flags should be consolidated into a struct.

pcalley replied:
done

msarvar commented on server/events/models/models.go:
```diff
@@ -411,6 +411,8 @@ type ProjectCommandContext struct {
 	// Verbose is true when the user would like verbose output.
 	Verbose bool
+	// Force is true when the apply should ignore apply_requirements.
+	Force bool
```
Let's name this ForceApply?

pcalley replied:
Renamed.

msarvar commented:
great job, looks good overall. Couple nits

nishkrishnan reviewed (commented):
Couple things. 

1.  Would be great if we could have a server flag that enables this since then we'd be able to move this upstream as well. Let's just do it as part of this PR since it should be pretty quick.
2. In the future, having the feature allocator be used in the least invasive way possible ensures you don't have to modify existing tests/ makes it easy to remove. Most obvious way to do this is using the decorator pattern.

```go
type SomeInterface interface {
  SomeFunc(...)
}


type SomeImpl struct {
}

func (s SomImpl) SomeFunc(...) {}

type AlternateImpl struct {}

func (s AlternateImpl) SomeFunc(...) {}

type FeatureAwareImpl struct{}

func (s FeatureAwareImpl) SomeFunc(...) {

   // call feature allocator and either route to AlternateImpl or SomeImpl

}
```

This makes it so that unit tests are scoped to the implementation and we can just remove the feature aware implementation once we're done without having to deal with fixing tests or editing implementation code.


In your case, one way to have done that would have been by only populating the `ProjectCommandContext` with the real value of force if the feature has been allocated and it could have been done as a delegate interface basically.

msarvar commented:
yeah, we should do what @nishkrishnan suggested and gate the force apply's value behind the feature flag decorator.

pcalley commented:
@nishkrishnan I'm definitely into the suggestion I'm not 100% I understand how to go about it. We have the https://github.com/lyft/atlantis/blob/da24e3c2646fdb5c38c2ce565d97ad7feb00b886/server/events/project_command_context_builder.go#L195 where we could move the FeatureAllocator logic and just pass the `force` value if the feature is enabled. I think that makes sense to me b/c it will simplify the project_command_runner.doApply() and then we wouldn't need to mock the feature allocator for the unit tests there. Is that sorta what you're thinking?

msarvar commented:
You would create a new decorator struct, similar to
https://github.com/lyft/atlantis/blob/da24e3c2646fdb5c38c2ce565d97ad7feb00b886/server/events/project_command_runner.go#L113.
It can have the feature allocator, you can then set the force apply flag to
false if the feature is not enabled.

On Thu, Sep 16, 2021 at 5:14 PM Paul Calley ***@***.***>
wrote:

> @nishkrishnan <https://github.com/nishkrishnan> I'm definitely into the
> suggestion I'm not 100% I understand how to go about it. We have the
> https://github.com/lyft/atlantis/blob/da24e3c2646fdb5c38c2ce565d97ad7feb00b886/server/events/project_command_context_builder.go#L195
> where we could move the FeatureAllocator logic and just pass the force
> value if the feature is enabled. I think that makes sense to me b/c it will
> simplify the project_command_runner.doApply() and then we wouldn't need to
> mock the feature allocator for the unit tests there. Is that sorta what
> you're thinking?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lyft/atlantis/pull/107#issuecomment-921349127>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAKSL5QJA4G4PS7BSOZ533UCKB7NANCNFSM5D65U66Q>
> .
>

pcalley commented on server/controllers/events/events_controller_e2e_test.go:
```diff
@@ -671,6 +674,7 @@ func setupE2E(t *testing.T, repoDir string) (events_controllers.VCSEventsControl
 	e2eGithubGetter := mocks.NewMockGithubPullGetter()
 	e2eGitlabGetter := mocks.NewMockGitlabMergeRequestGetter()
 	projectCmdOutputHandler := handlermocks.NewMockProjectCommandOutputHandler()
+	allocator := fmocks.NewMockAllocator()
```
SG. This got removed anyway from the changes above.

pcalley commented on server/events/project_command_runner.go:
```diff
@@ -395,6 +405,9 @@ func (p *DefaultProjectCommandRunner) doApply(ctx models.ProjectCommandContext)
 
+	if shouldAllocate && ctx.Force {
+		outputs = append(outputs, "WARNING: this apply run with --force option bypassing the apply requirements. This should only be used in an emergency.")
+	}
```
Cool I updated it so its a separate comment before the apply. Heres a screenshot. 
<img width="918" alt="Screen Shot 2021-09-17 at 4 09 21 PM" src="https://user-images.githubusercontent.com/1891649/133862752-543cb8b3-144a-4894-bd12-2cdf5478c4ea.png">

pcalley commented:
> Couple things.
> 
> 1. Would be great if we could have a server flag that enables this since then we'd be able to move this upstream as well. Let's just do it as part of this PR since it should be pretty quick.
> 2. In the future, having the feature allocator be used in the least invasive way possible ensures you don't have to modify existing tests/ makes it easy to remove. Most obvious way to do this is using the decorator pattern.
> 
> ```go
> type SomeInterface interface {
>   SomeFunc(...)
> }
> 
> 
> type SomeImpl struct {
> }
> 
> func (s SomImpl) SomeFunc(...) {}
> 
> type AlternateImpl struct {}
> 
> func (s AlternateImpl) SomeFunc(...) {}
> 
> type FeatureAwareImpl struct{}
> 
> func (s FeatureAwareImpl) SomeFunc(...) {
> 
>    // call feature allocator and either route to AlternateImpl or SomeImpl
> 
> }
> ```
> 
> This makes it so that unit tests are scoped to the implementation and we can just remove the feature aware implementation once we're done without having to deal with fixing tests or editing implementation code.
> 
> In your case, one way to have done that would have been by only populating the `ProjectCommandContext` with the real value of force if the feature has been allocated and it could have been done as a delegate interface basically.

Added the decorator logic here and updated the unit tests.
https://github.com/lyft/atlantis/blob/c4bf3c076faa6c3a3696b96449d3dfaf686fb476/server/events/project_command_runner.go#L156-L179

msarvar commented on server/events/project_command_runner.go:
```diff
@@ -152,6 +153,31 @@ func (p *ProjectOutputWrapper) updateProjectPRStatus(commandName models.CommandN
+	FeatureAllocator feature.Allocator
+}
+
+func (f *FeatureAwareProjectCommandRunner) Apply(ctx models.ProjectCommandContext) models.ProjectResult {
```
Can you add some  tests for this decorator as well?

pcalley replied:
Yeah totally just added some.

nishkrishnan commented on server/events/command_runner.go:
```diff
@@ -224,6 +224,13 @@ func (c *DefaultCommandRunner) RunCommentCommand(baseRepo models.Repo, maybeHead
 		ctx.Log.Err("Error running pre-workflow hooks %s. Proceeding with %s command.", err, cmd.Name.String())
 	}
 
+	if cmd.ForceApply {
+		warningMessage := "⚠️ WARNING ⚠️\n\n You have bypassed all apply requirements for this PR 🚀 . This can have unpredictable consequences 🙏🏽 and should only be used in an emergency 🆘 .\n\n 𝐓𝐡𝐢𝐬 𝐚𝐜𝐭𝐢𝐨𝐧 𝐰𝐢𝐥𝐥 𝐛𝐞 𝐚𝐮𝐝𝐢𝐭𝐞𝐝.\n"
+		if commentErr := c.VCSClient.CreateComment(baseRepo, pullNum, warningMessage, ""); commentErr != nil {
+			log.Err("unable to comment: %s", commentErr)
+		}
+	}
+
```
Shouldn't this be wrapped in the feature flag as well? The feature flag call should be idempotent and quick so you could theoretically employ the decorator pattern again here and wrap this.

pcalley replied:
Makes sense. I need to add some unit tests quick but how does this look:  https://github.com/lyft/atlantis/pull/107/commits/bb3607c6e55a6217966a1c7b9bbc367bbf85e363

pcalley replied:
done
